htlcswitch/switch.go

```diff
@@ -986,6 +996,25 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 			return s.handleLocalDispatch(packet)
 		}
 
+		// Before we attempt to find a non-strict forwarding path for
+		// this htlc, check whether the htlc is being routed over the
+		// same incoming and outgoing channel. If our node does not
+		// allow forwards of this nature, we fail the htlc early. This
+		// check is in place to disallow inefficiently routed htlcs from
+		// locking up our balance.
+		linkErr := checkCircularForward(
+			packet.incomingChanID, packet.outgoingChanID,
+			s.cfg.AllowCircularRoute, htlc.PaymentHash,
+		)
+		if linkErr != nil {
+			addErr := fmt.Errorf(circularRouteErrTemplate,
+				htlc.PaymentHash[:], packet.outgoingChanID)
+
+			return s.failAddPacket(
```

Comment:
I am wondering, what is the purpose of returning addErr here (via `failAddPacket`) and not just `linkErr`? `FailureDetail` already describes what is the problem.

Reply:
The `addrErr` provides the specific channel ID that the circle was attempted on. I've actually just removed this in #3844 though, I think we can get by without it. If we find ourselves wanting specific link errors, we can just make new `FailureDetail` types that have additional metadata.

Reply:
Going to kill `addrErr` here and just pass the `linkErr` in twice, since the change to remove `addErr` from `failAddPacket` is already in 3844.